### ISSUE TITLE
prefix not bound to a namespace for some attributes

### ISSUE THREAD

**Issue by f-christophe** [deegree/deegree3#600]
Hi,

There is an unbound attribute prefix in the output of a WFS request:

``` xml
<?xml version='1.0' encoding='UTF-8'?>
<wfs:FeatureCollection xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:schemaLocation="http://www.opengis.net/wfs/2.0 http://schemas.opengis.net/wfs/2.0/wfs.xsd http://www.opengis.net/gml/3.2 http://schemas.opengis.net/gml/3.2.1/gml.xsd urn:x-inspire:specification:gmlas:HydroPhysicalWaters:3.0 http://inspire.ec.europa.eu/schemas/hy-p/3.0/HydroPhysicalWaters.xsd" xmlns:wfs="http://www.opengis.net/wfs/2.0" timeStamp="2015-09-14T10:06:37Z" xmlns:gml="http://www.opengis.net/gml/3.2" numberMatched="unknown" numberReturned="0">
  <!--NOTE: numberReturned attribute should be 'unknown' as well, but this would not validate against the current version of the WFS 2.0 schema (change upcoming). See change request (CR 144): https://portal.opengeospatial.org/files?artifact_id=43925.-->
  <wfs:member>
    <hy-p:Watercourse xmlns:hy-p="urn:x-inspire:specification:gmlas:HydroPhysicalWaters:3.0" gml:id="HY-P_WATERCOURSE_f9fcb753-ffae-41ee-8cfc-62b7e1312c67">
      <gml:identifier codeSpace="http://inspire.jrc.ec.europa.eu/ids">http://inspire.ngi.be/so/BE/IGNB/ITGI-Vref/hy-p/Watercouse/HY_WatercourseSegment/{DBDA6CE1-C74E-490D-B74C-FFF7F408DE89}</gml:identifier>
      <hy-p:geographicalName xsi:nil="true" nilReason="other:unpopulated"/>
      <hy-p:hydroId xsi:nil="true" nilReason="other:unpopulated"/>
      <hy-p:beginLifespanVersion>1994-06-01T00:00:00</hy-p:beginLifespanVersion>
      <hy-p:geometry>
        <!--Inlined geometry 'HY-P_WATERCOURSE_f9fcb753-ffae-41ee-8cfc-62b7e1312c67_HY-P_GEOMETRY'-->
        <gml:LineString gml:id="HY-P_WATERCOURSE_f9fcb753-ffae-41ee-8cfc-62b7e1312c67_HY-P_GEOMETRY" srsName="EPSG:4326">
          <gml:posList>3.474799 51.184159 3.474824 51.184199 3.474841 51.184226</gml:posList>
        </gml:LineString>
      </hy-p:geometry>
      <hy-p:inspireId>
        <base:Identifier xmlns:base="urn:x-inspire:specification:gmlas:BaseTypes:3.2">
          <base:localId>hy-p.Watercouse.HY_WatercourseSegment.{DBDA6CE1-C74E-490D-B74C-FFF7F408DE89}</base:localId>
          <base:namespace>BE.IGNB.ITGI-Vref</base:namespace>
          <base:versionId xsi:nil="true" nilReason="other:unpopulated"/>
        </base:Identifier>
      </hy-p:inspireId>
      <hy-p:localType xsi:nil="true" gco:nilReason="other:unpopulated"/>
      <hy-p:origin xsi:nil="true" nilReason="other:unpopulated"/>
      <hy-p:persistence codeSpace="http://inspire.ec.europa.eu/codelist/HydrologicalPersistenceValue">perenial</hy-p:persistence>
      <hy-p:tidal xsi:nil="true" nilReason="other:unpopulated"/>
      <hy-p:bank xsi:nil="true"/>
      <hy-p:drainsBasin xsi:nil="true"/>
      <hy-p:neighbour xsi:nil="true"/>
      <hy-p:condition xsi:nil="true" nilReason="other:unpopulated"/>
      <hy-p:delineationKnown>true</hy-p:delineationKnown>
      <hy-p:length uom="meter">8.057386</hy-p:length>
      <hy-p:level>underground</hy-p:level>
      <hy-p:width xsi:nil="true" nilReason="unknown"/>
    </hy-p:Watercourse>
  </wfs:member>
</wfs:FeatureCollection>
```

More precisely here:

``` xml
      <hy-p:localType xsi:nil="true" gco:nilReason="other:unpopulated"/>
```

The gco prefix is not bound to a namespace. I think that this problem can be fixed by changing this line:

https://github.com/deegree/deegree3/blob/master/deegree-core/deegree-core-base/src/main/java/org/deegree/gml/feature/GMLFeatureWriter.java#L436

``` java
writeAttribute( writer, attr.getKey(), attr.getValue().getAsText() );
```

to this one:

``` java
writeAttributeWithNS( attr.getKey().getNamespaceURI(), attr.getKey().getLocalPart(), attr.getValue().getAsText() );
```

This method will write the missing namespace definition in case it doesn't exists.

Is this correct or am I missing something?

Thanks,
Christophe

**Comment by tfr42** [comment]
Can you please post which deegree version was used when reporting the bug? Can you please re-test your configuration with 3.4-RC1 and post your findings? Thanks in advance.

**Comment by f-christophe** [comment]
I was using version 3.3.17. I've rechecked with 3.4-RC1 but I have the same issue.

Changing [this](https://github.com/deegree/deegree3/blob/571d15524b2930ec2124f5ab440303797469702a/deegree-core/deegree-core-base/src/main/java/org/deegree/gml/feature/GMLFeatureWriter.java#L430) line:

`writeAttribute( writer, attr.getKey(), attr.getValue().getAsText() );`

to this one:

`writeAttributeWithNS( attr.getKey().getNamespaceURI(),attr.getKey().getLocalPart(), attr.getValue().getAsText() );`

fixes the problem for me but I don't know if it has side effects.

**Comment by tfr42** [comment]
@lgoltz what is your opinion?

**Comment by tfr42** [comment]
@f-christophe The related PR #1188 and #1135  have been merged into main branch for deegree v3.5-SNAPSHOT.
Please give it a try. In case of issues please open a new one.
